electroncash_gui/qt/lnsqt.py

```diff
@@ -0,0 +1,891 @@
+import time
+import requests
+from typing import Tuple, List, Callable
+from enum import IntEnum
+from electroncash import cashacct
+from electroncash import lns
+from electroncash import util
+from electroncash import web
+from electroncash.address import Address, UnknownAddress
+from electroncash.i18n import _, ngettext
+from electroncash.wallet import Abstract_Wallet
+
+
+class VerifyingDialog(WaitingDialog):
+
+    def __init__(self, parent, message, task, on_success=None, on_error=None, auto_cleanup=True,
+                 *, auto_show=True, auto_exec=False, title=None, disable_escape_key=False):
+        super().__init__(parent, message, task, on_success=on_success,
+                         on_error=on_error, auto_cleanup=auto_cleanup,
+                         auto_show=False, auto_exec=False,
+                         title=title or _('Verifying LNS Name'),
+                         disable_escape_key=disable_escape_key)
+        hbox = QHBoxLayout()
+        self._vbox.removeWidget(self._label)
+        icon_lbl = QLabel()
+        icon_lbl.setPixmap(QIcon(":icons/lns.png").pixmap(50))
+        hbox.addWidget(icon_lbl)
+        hbox.addWidget(self._label)
+        self._vbox.addLayout(hbox)
+        prog = QProgressBar()
+        prog.setRange(0,0)
+        self._vbox.addWidget(prog)
+        if auto_show and not auto_exec:
+            self.open()
+        elif auto_exec:
+            self.exec_()
+        destroyed_print_error(self)
+
+
+def verify_multiple_names(names : List[str], parent : MessageBoxMixin, wallet : Abstract_Wallet, timeout=10.0) -> int:
+    ''' Pass a list of names and will attempt to verify them all in 1 pass.
+    This is used by the Contacts tab to verify unverified LNS Names that
+    may have been imported. Returns the number of successfully verified names
+    or None on user cancel. '''
+    if not len(names):
+        return 0
+    names = set(names)
+    nnames = len(names)
+    q = queue.Queue()
+    def done_cb(thing):
+        if isinstance(thing, Exception):
+            raise thing
+        if isinstance(thing, List):
+            q.put(thing)
+        else:
+            q.put(None)
+    ctr = 0
+    def thread_func():
+        nonlocal ctr
+        wallet.lns.verify_name_asynch(name=names, success_cb=done_cb, error_cb=done_cb, timeout=timeout)
+        errs = 0
+        while ctr + errs < nnames:
+            try:
+                thing = q.get(timeout=timeout)
+                if thing is None:
+                    errs += 1
+                else:
+                    ctr += len(thing)
+            except queue.Empty:
+                return
+    code = VerifyingDialog(parent.top_level_window(),
+                           ngettext("Verifying {count} name please wait ...",
+                                    "Verifying {count} names please wait ...", nnames).format(count=nnames),
+                                    thread_func, auto_show=False, on_error=lambda e: parent.show_error(str(e))).exec_()
```

Comment:
This ended up touching the GUI from not-the-main-thread.  I fixed this in a commit that I will publish soon.

There are other errors like this and it led to random crashes in the error case.  FYI: The rule for Qt apps (and most GUI apps that interact directly with native windowing systems) is that you cannot touch the GUI from anything but the main thread.